qa/tasks/iscsi_gwcli.py

```diff
@@ -0,0 +1,180 @@
+        raise RuntimeError("gwcli {} failed!".format(check))
+
+def _run_ls(remote):
+    _run_check(remote)
+
+def _run_create(remote):
+    # create the datapool/blockX disk
+    args = ['gwcli', '/disks/', 'create', 'pool=datapool', 'image=blockX', 'size=1M']
+    remote.run(args=args)
+    check = 'blockX [.]+ \[datapool/blockX \(State.*\]'
+    _run_check(remote, check)
+
+    # create the iscsi-targets
+    target_iqn = "iqn.2003-01.com.redhat.iscsi-gw:ceph-gw"
+    args = ['gwcli', '/iscsi-targets', 'create', target_iqn]
+    remote.run(args=args)
+    check = '{} [.]+ \[Auth.*\]'.format(target_iqn)
+    _run_check(remote, check)
+
+    # the first gateway must be the local host
+    hostname = get_host()
+    ip0 = socket.gethostbyname(hostname)
+    args = ['gwcli', '/iscsi-targets/{}/gateways'.foramt(target_iqn), 'create', 'ip_addresses={}'.format(ip0), 'gateway_name={}'.format(hostname)]
+    remote.run(args=args)
+    check = '{0} [.]+ \[{1} \(UP\)\]'.format(hostname, ip0)
+    _run_check(remote, check)
+
+    # for all the other gateways
+    for ip in trust_ips:
+        if ip == ip0:
+            continue
+        hostname = get_host(ip)
+
+        args = ['gwcli', '/iscsi-targets/{}/gateways'.foramt(target_iqn), 'create', 'ip_addresses={}'.format(ip), 'gateway_name={}'.format(hostname)]
+        remote.run(args=args)
+        check = '{0} [.]+ \[{1} \(UP\)\]'.format(hostname, ip)
+        _run_check(remote, check)
+
+    # attach the disk to iscsi-target
+    client_iqn = "iqn.1994-05.com.redhat:client"
+    args = ['gwcli', '/iscsi-targets/{}/hosts'.format(target_iqn), 'create', client_iqn]
+    remote.run(args=args)
+    check = '{0} [.]+ \[Auth:.*\]'.format(client_iqn)
+    _run_check(remote, check)
+
+    # map the disk to the hosts
+    args = ['gwcli', '/iscsi-targets/{}/hosts/{}'.format(target_iqn, client_iqn), 'disk', 'disk=datapool/blockX']
+    remote.run(args=args)
+    check = 'lun 0 [.]+ \[datapool/blockX.*\]'
+    _run_check(remote, check)
+
+def _run_delete(remote):
+    # delete the hosts
+    args = ['gwcli', '/iscsi-targets/{}/hosts'.format(target_iqn), 'delete', 'client_iqn={}'.format(client_iqn)]
+    remote.run(args=args)
+    check = '{0} [.]+ \[Auth:.*\]'.format(client_iqn)
+    _run_check(remote, check, True)
+
+   # delete the disks
+    args = ['gwcli', '/iscsi-targets/{}/disks'.format(target_iqn), 'delete', 'disk=datapool/block02']
+    remote.run(args=args)
+    check = 'blockX [.]+ \[datapool/blockX \(State.*\]'
+    _run_check(remote, check, True)
+
+    # delete the iscsi-targets
+    args = ['gwcli', '/iscsi-targets', 'delete', target_iqn]
+    remote.run(args=args)
+    check = '{} [.]+ \[Auth.*\]'.format(target_iqn)
+    _run_check(remote, check, True)
+
+def _run_one_gateway(ctx, config, role):
+    """Spawned task that runs the gateway"""
+    role = config['gateways'][0]
+    (remote,) = ctx.cluster.only(role).remotes.keys()
+
+    # gwcli ls
+    _run_ls(remote)
```

Comment:
I would separate out the task for doing the baseline ceph-iscsi configuration from the actual tests. That would allow you to provide multiple tests with one common setup routing. i.e. you could create a workunit script that runs `nosetests` or similar to inject CLI arguments and verify the output from `gwcli`, another test could run block device IO tests via an initiator, another could exercise the REST API, etc.

Reply:
Yeah, sure.
For the first version I am planing to support the `gwcli` method only, which will make it simple and will handle all the other depending issues, like the teuthology fixings. After this for all the others like the REST API one will be easy.